[lukeai-tan] iP

lukeai-tan opened the pull request.

lukeai-tan committed:
Add constructor, and greet and farewell dialogues

- Added constructor to initialise Red Girl
- Added printBootSequence() to mimic Nier Automata's boot effect
- Added greet() to display a greeting message
- Added exit() to display a farewell message

lukeai-tan committed:
Add Echo functionality and improve existing code

- Added initRedGirl function
- Added more dialogue and with colour
- Renamed greet and exit functions to more intuitive names printGreeting and printFarewell
- Edited README to include Chatbot description

lukeai-tan committed:
Add addListEntry and printList functions

- printList function added, along with ArrayList to store strings to be printed
- addListEntry function takes in string input to be directly added to the list
- Adjusted Scanner to scan each line rather than each token, using sc.nextLine() instead of sc.next(), in order to work with addListEntry. This is done because each entry may consist of multiple words.
- Adopted the SE-EDU Java coding standard into my code

lukeai-tan committed:
Add Task class

Task contains a String description and a boolean isDone.
It has a constructor that takes in a String description parameter, as well as setting isDone to false by default.
getStatusIcon() creates a visual for its "done" status.
setAsDone() and setAsUndone are setter functions that sets the isDone variable to true and false respectively.
Overrode the default Object.toString() so that the Task instances displays completion status and task description / name.
Changed the ArrayList type in RedGirl from String to Task.

lukeai-tan committed:
Add convertToRedGirlTheme

Function that converts dialogue text colour into dark red.
Intention is to reduce redundancy in the print functions when the darkRed and reset strings are added to the front and back of the dialogue respectively.
This essentially makes all print functions less abstract, promoting SLAP principles

lukeai-tan committed:
Add markListEntry and unmarkListEntry

These two functions take in a task index and marks or unmarks the task as complete accordingly, printing a unique dialogue for each.

lukeai-tan committed:
Update RedGirl.java

Added redGirlPrint() that uses convertToRedGirlTheme(), which reduces more redundancy in the code when printing each of the bot's dialogues

Added handleMarkInput's functionality
- It handles edge cases such as wrong number of arguments and invalid index
- After edge case checks, it marks and unmarks accordingly

Modified initRedGirl's condition for the mark and unmark commands
- conditions now search for "mark " and "unmark " with a space behind, ensuring that it is the actual command and not a task named "marking homework" or "marker"

lukeai-tan committed:
Add parseInput()

Added parseInput to take in string inputs, reducing the abstraction level in initRedGirl

Fixed some coding standards that were missed

lukeai-tan committed:
Add support for Todo and Deadline tasks

Introduced TodoTask and DeadlineTask classes extending Task, which is now abstract.
Updated RedGirl to handle adding and displaying both todo and deadline tasks, including parsing input and printing appropriate messages.

lukeai-tan committed:
Add EventTask and refactor task handling

Introduced the EventTask class to represent events with start and end times
Updated RedGirl to handle event task input, refactored task addition logic to use a common addTask method, and improved user feedback messages
Also added a getTaskCount method and enhanced list printing for empty task lists

lukeai-tan committed:
Update EventTask.java

lukeai-tan committed:
Fix missed coding standards

lukeai-tan committed:
Update DeadlineTask.java

Added getter function for deadline

lukeai-tan committed:
Add todo input handler and default echo

Introduced handleTodoTaskInput to process 'todo' commands with validation.
Added echo method for unrecognized input, replacing the previous default behaviour of adding a todo entry.

lukeai-tan committed:
Add text ui testing

Added input and corresponding expected txt files to test for the correctness of the output of the commands todo, deadline, event and exit

lukeai-tan committed:
Rename chatbot to Red Girls

Changed the description of the README to fit the name change

lukeai-tan committed:
Add custom Red Girls exception class

Introduced RedGirlsException class to encapsulate custom error messages with themed output.
Also renamed the convertToRedGirlTheme to toRedGirlsString, which is more intuitive and reminiscent of the original toString function

lukeai-tan committed:
Refactor input handling to use exceptions

Replaces in-method error messages with custom RedGirlsException throws for invalid input cases in task handling methods.
Adds new exception types and updates method signatures to declare thrown exceptions, improving error handling and code clarity.

lukeai-tan committed:
Merge branch-Level-5

lukeai-tan committed:
Refactor package structure and add Parser

Moved Task-related classes to the 'purpose' package and RedGirlsException to the 'intellect' package.
Updated imports in RedGirls.java accordingly.
Added a new Parser class (empty for now) in the 'intellect' package to prepare for future parsing logic.

lukeai-tan committed:
Refactor task management into TaskList class

Moved all task-related logic from RedGirls.java into a new purpose.TaskList class to improve code organization and separation of concerns.
RedGirls now delegates task operations to TaskList, simplifying its responsibilities.

lukeai-tan committed:
Refactor input parsing logic into Parser class

Moved input parsing and task handling methods from RedGirls to a new Parser class for better separation of concerns. RedGirls now delegates input parsing to Parser, which manages task operations using the provided TaskList.

lukeai-tan committed:
Handle RedGirlsException within parseInput method

Wrapped the command parsing logic in a try-catch block to catch RedGirlsException and print its message, preventing exceptions from crashing the entire program every time.

lukeai-tan committed:
Refactor dialogue methods into Dialogue class

Moved all dialogue-related methods from RedGirls.java to a new charisma.Dialogue class for better separation of concerns and code organization.
Updated RedGirls to use the new Dialogue class.
Also removed unnecessary exception from Parser.parseInput signature.

lukeai-tan committed:
Refactor command parsing to use switch statement

Replaces multiple if-else statements with a switch-case structure for parsing user commands. This improves readability and maintainability of the parseInput method.

lukeai-tan committed:
Refactor TaskList to use static methods and fields

TaskList methods and the tasks list are now static, removing the need to instantiate TaskList.
Updated Parser and RedGirls to use the new static TaskList.
Added abstract Command class and a placeholder TodoCommand for future command pattern implementation.

lukeai-tan committed:
Add todo, deadline and event commands

Introduces TodoCommand, DeadlineCommand and EventCommand classes to handle creation and addition of todo, deadline and event tasks respectively.
Each command is now responsible for picking apart the inputs, creating the tasks and adding it to the static TaskList.

lukeai-tan committed:
Clean up old task input handling functions

Removed the old functions in Parser and TaskList that got replaced.

lukeai-tan committed:
Add MarkCommand for task mark and unmarking

Created MarkCommand class that extends Command to take the responsibility of handling mark input

lukeai-tan committed:
Move all packages into one main package

Created the main package redgirls which contains all the packages intellect, commands, charisma and purpose

lukeai-tan committed:
Add ListCommand

Created ListCommand for handling printing the task list

lukeai-tan committed:
Refactor command parsing

Parser's parseInput now returns a Command instead of executing the command itself, reducing its responsibility and promoting better SRP
Command is executed in RedGirls.java instead
Made the switch case more succinct using lamda style

lukeai-tan committed:
Merge branch-A-Packages

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Change Command execute function to return bool

Introduced ByeCommand to handle program termination, replacing direct 'bye' string checks in RedGirls.java.
Parser now handles the interpretation of all commands, while RedGirls.java only handles the execution of commands.
Updated Command abstract class to make execute function return a boolean. I originally intended to have ByeCommand call System.exit(0) to quit the app when the command was executed. However, I found making the execute() a boolean return type to be checked by RedGirls.java cleaner.

lukeai-tan committed:
Add HelpCommand to display command list

Added HelpCommand to list all the commands and their descriptions

hyeungac commented on src/main/java/redgirls/charisma/Dialogue.java:
```diff
@@ -0,0 +1,76 @@
+        redGirlsPrint("Our exchange concludes. Your thoughts linger...as do mine.");
+    }
+
+    public static void printCommandList() {
```
good job bro, listing out all of the functions of your bot

lukeai-tan replied:
Thanks!

hyeungac commented on src/main/java/redgirls/purpose/DeadlineTask.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Override
+    public String toString() {
+        return "[D]" + super.toString() + " (by: " + getDeadline() + ")";
```
do you need parenthesis here? im not sure..

hyeungac commented on src/main/java/redgirls/RedGirls.java:
```diff
@@ -0,0 +1,33 @@
+package redgirls;
```
interesting package here

lukeai-tan replied:
I agree. It is truly a package.

hyeungac commented on src/main/java/redgirls/intellect/Parser.java:
```diff
@@ -0,0 +1,29 @@
+            throw RedGirlsException.emptyInput();
+        }
+        String command = input.split(" ")[0];
+        return switch (command) {
```
nice switch skill...

lukeai-tan replied:
Thanks! IntelliJ gave me a suggestion to use that approach instead. Praise the IntelliJ

blimc1 commented on src/main/java/redgirls/charisma/Dialogue.java:
```diff
@@ -0,0 +1,76 @@
+        for (String line : bootLines) {
+            System.out.println(line);
+            try {
+                Thread.sleep(200);
```
Could use a constant for the 200, like BOOT_DELAY. Gives it a bit more meaning.

lukeai-tan replied:
Thanks for the idea! I will change this.

lukeai-tan committed:
Rename package from redgirls to n2

All Java source files have been moved from the 'redgirls' package to the 'n2' package

blimc1 commented on src/main/java/redgirls/purpose/TaskList.java:
```diff
@@ -0,0 +1,77 @@
+
+import java.util.ArrayList;
+
+public class TaskList {
```
Nice naming of the classes in general. Very informative.

blimc1 commented on src/main/java/redgirls/purpose/TaskList.java:
```diff
@@ -0,0 +1,77 @@
+        }
+    }
+
+    public static void addTodo(TodoTask task) {
```
Nice use of verbs for the methods

blimc1 commented on src/main/java/redgirls/RedGirls.java:
```diff
@@ -0,0 +1,33 @@
+    public static void initRedGirl() {
+        Dialogue.printBootSequence();
+        Dialogue.printGreeting();
+        try (Scanner sc = new Scanner(System.in)) {
```
Could have a more descriptive name for the scanner, but I'm just nit picking here

blimc1 reviewed (commented):
Nothing much to comment on to be honest

lukeai-tan committed:
Refactor boot delay to use constant

Introduced a static final BOOT_DELAY_MS constant to replace the hardcoded boot delay value in the Dialogue class. This improves maintainability and makes it easier to adjust the delay in the future.

lukeai-tan committed:
Update TaskList.java

lukeai-tan committed:
Refactor TaskList to unify task addition methods

Replaces addTodo, addDeadline, and addEvent with a single public addTask method in TaskList.
Updates command classes to use addTask, and renames getTaskCount to size to follow Java naming conventions

lukeai-tan committed:
Add input length check for TodoCommand

Added a check to ensure the input length is greater than 5 before processing a todo task. This prevents invalid todo tasks with insufficient input from being accepted.

lukeai-tan committed:
Add typing effect to redGirlsPrint output

Modified redGirlsPrint to display dialogue with a typing effect by introducing the typingEffectPrint method. This enhances the user experience by simulating text being typed out.

lukeai-tan committed:
Add and implement task deletion

Added DeleteCommand that executes deleteTask
deleteTask gets the task at the specified index and removes it from the ArrayList tasks

lukeai-tan committed:
Add and implement task deletion

Added DeleteCommand that executes deleteTask
deleteTask gets the task at the specified index and removes it from the ArrayList tasks

lukeai-tan committed:
Add MemoryArchive class to save tasks

Added a new class MemoryArchive for saving and loading tasks in memory
Added a serialize abstract method in Task to ensure all tasks implement it
Serialize converts a task into a formatted string
Added a new RedGirlsException for when the tasks cannot be written to the specified file path

lukeai-tan committed:
Add deserialize and load method to MemoryArchive

Added a deserialize method to parse the serialized task strings
Added a load method that uses serialize to convert a save file into a TaskList

lukeai-tan committed:
Include save function in each task related command

When tasks are manipulated in any way using todo, deadline, event, mark, unmark, the task list is saved.

lukeai-tan committed:
Update MemoryArchive.java

Changed filepath to the root
Added edge case where if parent directory "file" was not created or found, it would be created

lukeai-tan committed:
Update .gitignore

lukeai-tan committed:
Revert "Add and implement task deletion"

This reverts commit 2d84eb0f65d94200f21809de0b872fc275a9d43e.

lukeai-tan committed:
Merge branch-Level-6 into master

lukeai-tan committed:
Reapply "Add and implement task deletion"

This reverts commit 2a8a86e05117d4a99bb00b09c6e33d03c1d9ccbc.

lukeai-tan committed:
Revert "Reapply "Add and implement task deletion""

This reverts commit 7579269b753e106acb0bed073f511f32bd3bafea.

lukeai-tan committed:
Add and implement task deletion

Added DeleteCommand that executes deleteTask
deleteTask gets the task at the specified index and removes it from the ArrayList tasks

lukeai-tan committed:
Merge branch-Level-6 into master

lukeai-tan committed:
Merge branch-Level-7 into master

lukeai-tan committed:
Edit dialogue related to deleting tasks

lukeai-tan committed:
Make exception messages more descriptive

Added a line parameter to the unknownTaskType exception to display the corrupted task when deserializing it upon loading the save file

lukeai-tan committed:
Update .gitignore

lukeai-tan committed:
Edit gitignore

lukeai-tan committed:
Update .gitignore

lukeai-tan committed:
Update TaskList.java

lukeai-tan committed:
Update .gitignore

lukeai-tan committed:
Update the saving frequency

Added MemoryArchive.save() to the main while loop in the main file, where the task list gets saved every time a line gets parsed

lukeai-tan committed:
Update Dialogue.java

Added the delete description in help

NigelYeoTW commented on src/main/java/n2/RedGirls.java:
```diff
@@ -0,0 +1,36 @@
+        Dialogue.printBootSequence();
+        Dialogue.printGreeting();
+        try (Scanner sc = new Scanner(System.in)) {
+            Parser parser = new Parser();
+            while (true) {
+                String s = sc.nextLine();
+                try {
+                    Command c = parser.parseInput(s);
+                    if (!c.execute()) {
+                        break;
+                    }
+                    MemoryArchive.save(TaskList.getTaskList());
+                } catch (RedGirlsException e) {
+                    System.out.println(e.getMessage());
+                }
+            }
+        }
+    }
```
Perhaps extract out functions to reduce nesting

NigelYeoTW commented on src/main/java/n2/RedGirls.java:
```diff
@@ -0,0 +1,36 @@
+    public static void initRedGirl() {
+        Dialogue.printBootSequence();
+        Dialogue.printGreeting();
+        try (Scanner sc = new Scanner(System.in)) {
```
Should there be a corresponding catch block?

NigelYeoTW commented on src/main/java/n2/charisma/Dialogue.java:
```diff
@@ -0,0 +1,96 @@
+
+    public static void redGirlsPrint(String dialogue) {
+        System.out.println();
+        typingEffectPrint(toRedGirlsString(dialogue), 5);
```
Perhaps reduce the magic numbers for ease of reading, this is seen in other areas of the code as well

NigelYeoTW commented on src/main/java/n2/commands/ByeCommand.java:
```diff
@@ -0,0 +1,11 @@
+
+public class ByeCommand extends Command {
+    @Override
+    public boolean execute() {
```
Perhaps the function could sound more like a boolean, this issue is seen is other areas of the code as well

NigelYeoTW commented on src/main/java/n2/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,42 @@
+    public void handleDeleteTaskInput(String input) throws RedGirlsException {
+        String[] parts = input.split("\\s+");
+        if (parts.length != 2) {
+            if(parts.length < 2) {
```
Perhaps have consistent whitespace for ease of reading

NigelYeoTW commented on src/main/java/n2/commands/DeleteCommand.java:
```diff
@@ -0,0 +1,42 @@
+            if(parts.length < 2) {
+                throw RedGirlsException.invalidDelete();
+            } else {
+                throw RedGirlsException.invalidTaskIndex();
+            }
+        }
```
Perhaps introduce guard clauses to reduce nesting

NigelYeoTW commented on src/main/java/n2/commands/EventCommand.java:
```diff
@@ -0,0 +1,39 @@
+            eventTask = new EventTask(description, from, to);
+        } else {
+            throw RedGirlsException.missingEventEndTime();
+        }
```
Perhaps make the happy path more prominent?

NigelYeoTW commented on src/main/java/n2/commands/MarkCommand.java:
```diff
@@ -0,0 +1,50 @@
+            }
+        }
+
+        this.command = parts[0];
```
Perhaps reduce the usage of 'this' when not needed to remove redundancy

lukeai-tan replied:
Thanks for catching this!

NigelYeoTW commented on src/main/java/n2/intellect/Parser.java:
```diff
@@ -0,0 +1,31 @@
+        if (input == null || input.isBlank()) {
+            throw RedGirlsException.emptyInput();
+        }
+        String command = input.split(" ")[0];
+        return switch (command) {
+            case "list" -> new ListCommand(input);
+            case "mark", "unmark" -> new MarkCommand(input);
+            case "deadline" -> new DeadlineCommand(input);
+            case "event" -> new EventCommand(input);
+            case "todo" -> new TodoCommand(input);
+            case "delete" -> new DeleteCommand(input);
+            case "bye" -> new ByeCommand();
+            case "help" -> new HelpCommand();
+            default -> throw RedGirlsException.unknownCommand();
```
Perhaps standardize your switch case indentation for your lambda style switch statements as there seems to be another block that does not have indentation

lukeai-tan replied:
Good catch. I will ammend this.

NigelYeoTW commented on src/main/java/n2/memory/MemoryArchive.java:
```diff
@@ -0,0 +1,93 @@
+        case "D":
+            if (parts.length < 4) {
+                throw RedGirlsException.invalidDeadlineTask();
+            }
+            String deadline = parts[3].trim();
+            return new DeadlineTask(description, deadline, isDone);
+
+        case "E":
+            if (parts.length < 4) {
+                throw RedGirlsException.invalidEventTask();
+            }
+            String[] times = parts[3].trim().split("-");
+
+            if (times.length < 2) {
+                throw RedGirlsException.invalidEventTask();
+            }
+            String from = times[0].trim();
+            String to = times[1].trim();
+            return new EventTask(description, from, to, isDone);
+
+        default:
+            throw RedGirlsException.unknownTaskType(line);
+        }
+    }
+
+    public static ArrayList<Task> load() throws RedGirlsException {
```
Perhaps variable and function naming could be more intuitive

NigelYeoTW commented on src/main/java/n2/purpose/TaskList.java:
```diff
@@ -0,0 +1,97 @@
+        }
+    }
+
+    public static int size() {
```
Perhaps function name could be more intuitive?

NigelYeoTW reviewed (commented):
Overall good effort, however, there seems to be some minor coding standard issues as well as consistency issue through the code, do look though the code as some of these issues are seen repeating throughout. Keep it up! 🔥

lukeai-tan committed:
Improve code quality and follow standards

Removed magic numbers in Dialogue and replaced it with a more intuitive name
Removed nesting in DeleteCommand for the parts.length check
Removed unnecessary this. usages in the code

lukeai-tan committed:
Refactor the execute method

Changed all the commands' execute method back into a void return type. Instead of using execute to return a bool for whether the command leads to an exit of the program, a separate bool return type method isExit() is used .

lukeai-tan committed:
Add DateConverter class

Created DateConverter class that parses both Date and DateTime strings into a specific format
It recognises a variety of date time patterns

lukeai-tan committed:
Add DateTime parsing to Event and Deadline

lukeai-tan committed:
Update DateConverter.java

Added more date time formats as well as making the AM and PM case insensitive in parseDateTime

lukeai-tan committed:
Revert "Update DateConverter.java"

This reverts commit 86adeecd9ffaf763724176e1da1cdd3521856911.

lukeai-tan committed:
Fix branch issues

lukeai-tan committed:
Add DateConverter class

Created DateConverter class that parses both Date and DateTime strings into a specific format
It recognises a variety of date time patterns

lukeai-tan committed:
Merge pull request #2 from lukeai-tan/branch-Level-8

Merge branch Level 8

lukeai-tan committed:
Abstract out date time formatting

Created handleDateTimeParsing to simplify the responsibilities of each Command

lukeai-tan committed:
Add FindCommand

Created FindCommand that identifies the keyword and calls a function printFilteredList in TaskList, which prints out all the tasks with said keyword
If keyword is empty or white space, exception is called

lukeai-tan committed:
Add FindCommand to parser

Added the find command to the parser, and also edited TaskList's printFilteredList to only find the keyword in the task description, not the whole task string itself.

lukeai-tan committed:
Merge pull request #3 from lukeai-tan/branch-Level-9

Branch level 9

lukeai-tan committed:
Add JavaDoc to TaskList

lukeai-tan committed:
JavaDoc for TodoTask class

lukeai-tan committed:
Add JavaDoc for Task class

lukeai-tan committed:
Add JavaDoc for EventTask

lukeai-tan committed:
Add JavaDoc for DeadlineTask

lukeai-tan committed:
Add JavaDoc to MemoryArchive

lukeai-tan committed:
Add JavaDoc to Parser

lukeai-tan committed:
Add JavaDoc for Command

lukeai-tan committed:
Add JavaDoc for ByeCommand and HelpCommand

lukeai-tan committed:
Add JavaDoc to FindCommand

lukeai-tan committed:
Add JavaDoc to ListCommand

lukeai-tan committed:
Add JavaDoc to TodoCommand

Removed unnecessary RedGirlException throws for all task related commands

lukeai-tan committed:
Add JavaDoc to RedGirlsException and Commands

lukeai-tan committed:
Add JavaDoc to Dialogue

lukeai-tan committed:
Fix bugs in DeadlineCommand and EventCommand

Fixed issue where null checks did nothing and crashed the program when parameters set as null were set as parameters.
Also updated the text UI testing inputs and EXPECTED.TXT

lukeai-tan committed:
Add JavaDoc to DateConverter

lukeai-tan committed:
Add JavaDoc to RedGirls main

lukeai-tan committed:
Add command usage in JavaDoc

lukeai-tan committed:
Merge pull request #4 from lukeai-tan/branch-A-JavaDoc

Branch a java doc

lukeai-tan committed:
Create User Guide

Added intro and feature list to user guide

lukeai-tan committed:
Add specified exceptions for Event Command

Updated test cases to fit the changes

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Create _config.yml

lukeai-tan committed:
Delete _config.yml

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Fix bug for mark command

lukeai-tan committed:
Update ui text test cases

lukeai-tan committed:
Promote SLAP in EventCommand

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Edit README.md with lore

lukeai-tan committed:
Fix coding quality issues

lukeai-tan committed:
Update DateConverter.java

lukeai-tan committed:
Update README.md

lukeai-tan committed:
Fix EventCommand bug

lukeai-tan committed:
Update DeleteCommand.java

lukeai-tan committed:
Update DeleteCommand.java
